engine/src/main/java/org/hibernate/search/backend/impl/InTransactionWorkQueueSynchronization.java

```diff
@@ -0,0 +1,133 @@
+	 * (direct or indirect) reference to the Session.
+	 */
+
+	private final QueueingProcessor queueingProcessor;
+	private boolean consumed;
+	private boolean prepared;
+	private final ConcurrentMap<Object, WorkQueueSynchronization> queuePerTransaction;
+	private final WorkQueue queue;
+	private final Object transactionIdentifier;
+
+	/**
+	 * in transaction work
+	 */
+	public InTransactionWorkQueueSynchronization(Object transactionIdentifier, QueueingProcessor queueingProcessor,
+												ConcurrentMap<Object, WorkQueueSynchronization> queuePerTransaction,
+												ExtendedSearchIntegrator extendedIntegrator) {
+		this.transactionIdentifier = transactionIdentifier;
+		this.queueingProcessor = queueingProcessor;
+		this.queuePerTransaction = queuePerTransaction;
+		queue = new WorkQueue( extendedIntegrator );
+	}
+
+	public void add(Work work) {
+		queueingProcessor.add( work, queue );
+	}
+
+	public boolean isConsumed() {
+		return consumed;
+	}
+
+	@Override
+	public void beforeCompletion() {
+		// we are doing all the work in the before completion phase so that it is part of the transaction
+		try {
+			if ( prepared ) {
+				if ( log.isTraceEnabled() ) {
+					log.tracef(
+							"Transaction's beforeCompletion() phase already been processed, ignoring: %s", this.toString()
+					);
+				}
+			}
+			else {
+				if ( log.isTraceEnabled() ) {
+					log.tracef( "Processing Transaction's beforeCompletion() phase: %s", this.toString() );
+				}
+				queueingProcessor.prepareWorks( queue );
+				prepared = true;
+			}
+
+			if ( log.isTraceEnabled() ) {
+				log.tracef(
+						"Processing Transaction's afterCompletion() phase for %s. Performing work.", this.toString()
+				);
+			}
+			queueingProcessor.performWorks( queue );
+		}
+		finally {
+			consumed = true;
+			//clean the Synchronization per Transaction
+			//not needed stricto sensus but a cleaner approach and faster than the GC
+			if ( queuePerTransaction != null ) {
+				queuePerTransaction.remove( transactionIdentifier );
+			}
+		}
+	}
+
+	@Override
+	public void afterCompletion(int i) {
+		try {
+			if ( Status.STATUS_COMMITTED == i ) {
+				if ( log.isTraceEnabled() ) {
+					log.tracef(
+							"Processing Transaction's afterCompletion() phase for %s. Performing work.", this.toString()
+					);
+				}
+				queueingProcessor.performWorks( queue );
+			}
+			else {
+				if ( log.isTraceEnabled() ) {
+					log.tracef(
+							"Processing Transaction's afterCompletion() phase for %s. Cancelling work due to transaction status %d",
+							this.toString(),
+							i
+					);
+				}
+				queueingProcessor.cancelWorks( queue );
+			}
+		}
+		finally {
+			consumed = true;
+			//clean the Synchronization per Transaction
+			//not needed stricto sensus but a cleaner approach and faster than the GC
+			if ( queuePerTransaction != null ) {
+				queuePerTransaction.remove( transactionIdentifier );
+			}
+		}
+	}
+
+	public void flushWorks() {
+		WorkQueue subQueue = queue.splitQueue();
```

Comment:
this {{flushWork}} method is used by the explicit {{ftsession.flushToIndexes()}}, does your questions still stand?